### ISSUE TITLE
`constexpr size_to_sizeclass` has a significant performance impact

### ISSUE THREAD

**Issue by pentelbart** [microsoft/snmalloc#584]
I haven't tested this on other platforms, only on CHERI RISC-V

The change to `constexpr` (#563) for `size_to_sizeclass` required `sizeclass_lookup` to lose `static` ([here](https://github.com/microsoft/snmalloc/blob/38d4483b2797d1c00eff9326f24b5cd609f52b4f/src/snmalloc/mem/sizeclasstable.h#L414)). My compiler emits this as a memcpy of the whole `SizeClassLookup` struct which runs every time the function is called, making it really costly for small allocations (~13k more instructions) . Changing it back eliminates the overhead, though I'm not clear enough on the surrounding context to know if this is the right move.

**Comment by mjp41** [comment]
On x86 with Clang-10, the fast path for alloc is 
```asm
0000000000001bc0 <operator new(unsigned long)>:
    1bc0:       48 8d 47 ff             lea    -0x1(%rdi),%rax
    1bc4:       48 3d ff df 00 00       cmp    $0xdfff,%rax
    1bca:       77 3a                   ja     1c06 <operator new(unsigned long)+0x46>
    1bcc:       48 c1 e8 04             shr    $0x4,%rax
    1bd0:       48 8d 0d 41 96 00 00    lea    0x9641(%rip),%rcx        # b218 <snmalloc::sizeclass_metadata+0x1
208>
    1bd7:       0f b6 14 08             movzbl (%rax,%rcx,1),%edx
    1bdb:       48 8b 35 06 f4 00 00    mov    0xf406(%rip),%rsi        # 10fe8 <.got+0x18>
    1be2:       64 48 8b 04 25 00 00    mov    %fs:0x0,%rax
    1be9:       00 00
    1beb:       48 01 f0                add    %rsi,%rax
    1bee:       48 8d 0c d0             lea    (%rax,%rdx,8),%rcx
    1bf2:       64 48 8b 04 d6          mov    %fs:(%rsi,%rdx,8),%rax
    1bf7:       48 85 c0                test   %rax,%rax
    1bfa:       74 22                   je     1c1e <operator new(unsigned long)+0x5e>
    1bfc:       48 8b 10                mov    (%rax),%rdx
    1bff:       0f 18 0a                prefetcht0 (%rdx)
    1c02:       48 89 11                mov    %rdx,(%rcx)
    1c05:       c3                      retq
...
```

Is this where you are seeing the explosion in instruction on RISV?

**Comment by mjp41** [comment]
This was from looking at  `objdump -d libsnmallocshim.so -C`.

**Comment by mjp41** [comment]
My understanding was that `constexpr` on a field implies both `static` and `inline`.

Was it replacing the `static inline` on the `sizeclass_lookup` that fixed the problem, or on the function call?

**Comment by mjp41** [comment]
Also, is this a Debug build?

**Comment by pentelbart** [comment]
> Was it replacing the `static inline` on the `sizeclass_lookup` that fixed the problem, or on the function call?

Both, because `sizeclass_lookup` can't be `static` when the function is `constexpr`.


> Also, is this a Debug build?

I think so; it has debug symbols at least.

**Comment by mjp41** [comment]
Can you increase the optimisation level and see if it is still like this?

**Comment by mjp41** [comment]
So on Linux Debug, I am seeing:
```asm
000000000000f0b0 <snmalloc::size_to_sizeclass(unsigned long)>:
    f0b0:       48 81 ec 28 10 00 00    sub    $0x1028,%rsp
    f0b7:       48 89 bc 24 18 10 00    mov    %rdi,0x1018(%rsp)
    f0be:       00
    f0bf:       48 c7 84 24 10 10 00    movq   $0xfff,0x1010(%rsp)
    f0c6:       00 ff 0f 00 00
    f0cb:       48 8d 44 24 10          lea    0x10(%rsp),%rax
    f0d0:       48 89 c7                mov    %rax,%rdi
    f0d3:       48 8d 35 a6 48 02 00    lea    0x248a6(%rip),%rsi        # 33980 <snmalloc::sizeclass_metadata+0
x1208>
    f0da:       ba ff 0f 00 00          mov    $0xfff,%edx
    f0df:       e8 1c 75 ff ff          callq  6600 <memcpy@plt>
    f0e4:       48 8b bc 24 18 10 00    mov    0x1018(%rsp),%rdi
    f0eb:       00
    f0ec:       e8 3f 02 00 00          callq  f330 <snmalloc::sizeclass_lookup_index(unsigned long)>
    f0f1:       48 89 44 24 08          mov    %rax,0x8(%rsp)
    f0f6:       48 81 7c 24 08 ff 0f    cmpq   $0xfff,0x8(%rsp)
```

So `f0df` looks suspiciously like what you are seeing.

**Comment by mjp41** [comment]
I think it is copying the `sizeclass_metadata` table on to the stack.  I never perf test Debug builds, but I can see this is not ideal.

**Comment by mjp41** [comment]
Could you try #585 and see if it addresses the issue for you.

**Comment by pentelbart** [comment]
Yes, that seems to work!

**Comment by mjp41** [comment]
Closed by #585
